### PR TITLE
[JENKINS-68542] Store inbound cookie key in connection state

### PR DESCRIPTION
**Pull request by Vlatombe**
This ends up being stored in the channel properties, so that it can be
checked later on if the agent reconnects while the controller still
thinks it is connected.

See [JENKINS-68542](https://issues.jenkins.io/browse/JENKINS-68542).

### Proposed changelog entries

* Fix websocket reconnection in edge cases.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

### CODE REVIEW

**Review by jglick — APPROVED** [review]
Please pick up the `remoting` PR via incremental dep so that end-to-end tests of WS agents run with both parts of the fix.

File: core/src/main/java/jenkins/agents/WebSocketAgents.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/agents/WebSocketAgents.java:99]: Spotbugs raising `HTTP_RESPONSE_SPLITTING`

**Comment by jglick** [comment]
Ready for review?

**Review by jglick — DISMISSED** [review]

File: core/src/main/java/jenkins/websocket/WebSocketSession.java
Comment (jglick) [inline core/src/main/java/jenkins/websocket/WebSocketSession.java:139]: Cleaner generally, though I would still recommend patching `SlaveComputer` to catch any `Exception`, just in case.

**Review by jglick — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/jenkins/websocket/WebSocketSession.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/websocket/WebSocketSession.java:139]: Done

**Review by jglick — APPROVED** [review]

**Comment by Vlatombe** [comment]
Please hold on the merge.

**Comment by Vlatombe** [comment]
I'm hitting an issue with
https://github.com/jenkinsci/remoting/blob/352268273d3beb2808a3c5b29022dad8b305312b/src/main/java/org/jenkinsci/remoting/engine/JnlpConnectionState.java#L69
and using nginx as a reverse proxy. It seems to filter such header names.

**Comment by jglick** [comment]
I suspect dots (`.`) are being blocked in HTTP header names. https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2 does not say these should be considered special characters in this context but who knows? Anyway using something named according to `JnlpAgentProtocol` in a WebSocket header name is odd. Maybe switch to another header name for WebSocket transport, like `Connection-Cookie`? There should be no compatibility issue since no shipping version of Remoting interprets the current `COOKIE_KEY` anyway, right?

**Comment by Vlatombe** [comment]
When not using websockets, it might.

File: core/src/main/java/jenkins/agents/WebSocketAgents.java
Comment (jglick) [inline core/src/main/java/jenkins/agents/WebSocketAgents.java:96]: Maybe call it `…_HEADER` for clarity?

**Review by jglick — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by alecharp — APPROVED** [review]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
